flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java

```diff
@@ -133,9 +138,18 @@ private String getJobClassNameOrScanClassPath() throws FlinkException {
 	}
 
 	private String scanClassPathForJobJar() throws IOException {
-		LOG.info("Scanning class path for job JAR");
-		JarFileWithEntryClass jobJar = JarManifestParser.findOnlyEntryClass(jarsOnClassPath.get());
-
+		final JarFileWithEntryClass jobJar;
+		if (getUserClassPaths().isEmpty()) {
+			LOG.info("Scanning system class path for job JAR");
+			jobJar = JarManifestParser.findOnlyEntryClass(jarsOnClassPath.get());
+		} else {
+			LOG.info("Scanning user class path for job JAR");
+			final List<File> userJars = getUserClassPaths()
+				.stream()
+				.map(url -> new File(url.getFile()))
+				.collect(Collectors.toList());
+			jobJar = JarManifestParser.findOnlyEntryClass(userJars);
+		}
```

Comment:
Do you mean that we should search for an entry class both in the user classpath and system classpath?
I have a little concern about that. If there is only an entry class in the system class and we use it. I think it leads to the jars in the job directory will not be used. It is a little weird that the user provides some jars but could not be used.

Reply:
Wouldn't we still use the jars from the `usrlib`? Only the job jar containing the manifest would come from the `lib` directory.

I find it a bit weird that sometimes we only look in one directory and some other times on the system class path.

Alternatively,  I would remove support for looking for job jars in the system class path and rename the retriever to something like `UsrLibJobGraphRetriever`.

Reply:
> Alternatively, I would remove support for looking for job jars in the system classpath and rename the retriever to something like UsrLibJobGraphRetriever.

It's an option. There might be some of the situation that the user himself builds the image.
So I think we might introduce two classes `UsrLibJobGraphRetriever` and `SysLibJobGraphRetriever`.  One is to retrieve the `jobgraph` from `usrlib`. The other is to retrieve the `jobgraph` from `lib`.

We could introduce a ClassPathJobGraphRetrieverFactory to create the JobGraphRetriever according to whether the `usrlib` exsits or not.

What do you think?

Reply:
Hmm I still don't fully understand why it would be a problem to first look into `usrlib` and then `lib` for a jar with a configured entrypoint class. What exactly are your concerns with this approach @guoweiM?

Reply:
If the user provides an entrypoint class there is no problem. Because this method `scanClassPathForJobJar` would not be called.
However, if the user does not provide an entrypoint and we need to find an entrypoint class there is a problem.

For example:
usrlib/
        |_xxx.jar #no entry point class
lib/
    |_yyy.jar #entry point class

If we find an entrypoint class from yyy.jar, I assume that this entry point class would be load by the parent classloader. This also means that xxx.jar would not be loaded in any situation.
The user configures the `usrlib` and no jar in this directory would be used. I think we should avoid this situation and notify the user early.

Reply:
Ok, your concern is that the user misconfigured the setup and expects that something from `xxx.jar` will be loaded even though the jar with the entrypoint is in `lib`.

If we say that the user jar must be in `usrlib` and it fails if he puts it in `lib` then we get an inconsistency. The problem is that the user can still put his user jar in `lib` if he specifies the entrypoint class name. This is then a somewhat inconsistent behaviour which is weird. In that sense, it would be more consistent to say that Flink uses either the system or user code classloader depending where the user code jar is/can be found.

Reply:
If the user specifies an entrypoint class name, we could check whether the entrypoing class can be found in the `usrlib` and throw an exception if it can not be found. 
From the perspective of the user,  I think this could give the user a more clear expectation. The entrypoint should be in the usrlib if you configure it.

Reply:
This could be a solution to the inconsistency. Is there any scenario where the user would like to have the user code explicitly in `lib`?

Reply:
Then let's do it like this :-)

Reply:
thanks!